src/main/host/descriptor/mod.rs

```diff
@@ -203,54 +203,61 @@ impl StatusEventSource {
 }
 
 /// Represents a POSIX description, or a Linux "struct file".
+#[derive(Clone)]
 pub enum PosixFile {
-    Pipe(pipe::PipeFile),
+    Pipe(Arc<AtomicRefCell<pipe::PipeFile>>),
 }
 
 // will not compile if `PosixFile` is not Send + Sync
 impl IsSend for PosixFile {}
 impl IsSync for PosixFile {}
 
 impl PosixFile {
-    pub fn read(&mut self, bytes: &mut [u8], event_queue: &mut EventQueue) -> SyscallReturn {
+    pub fn read(&self, bytes: &mut [u8], event_queue: &mut EventQueue) -> SyscallReturn {
         match self {
-            Self::Pipe(f) => f.read(bytes, event_queue),
+            Self::Pipe(f) => f.borrow_mut().read(bytes, event_queue),
```

Comment:
I modified the structure a bit in a second commit so that we don't lose this flexibility, but it requires some messy boilerplate code. Does it look reasonable to you, and do you have any ideas of ways to make it cleaner?

With the changes, some example code would look like the following:

```rust
pub fn my_func(f: PosixFile) {
    // can pattern match at the top level if you need the Arc
    if let PosixFile::Pipe(ref p) = f {
        p.borrow().status();
    }

    let mut f = f.borrow_mut();

    // can call functions directly on the borrowed reference
    let buf = [0u8; 10];
    EventQueue::queue_and_run(|event_queue| {
        f.write(&buf, event_queue);
    });

    // can pattern match on the borrowed reference
    if let PosixFileRefMut::Pipe(p) = f {
        p.status();
    }
}
```